File: app/src/main/java/com/pr0gramm/app/orm/CachedVote.java
package com.pr0gramm.app.orm;

import android.annotation.TargetApi;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;

import com.google.common.base.Joiner;
import com.google.common.base.Optional;
import com.google.common.collect.Ordering;
import com.pr0gramm.app.feed.Vote;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import static com.google.common.collect.Lists.transform;

/**
 */
public class CachedVote {
    private static final Logger logger = LoggerFactory.getLogger("CachedVote");

    public final long itemId;
    public final Type type;
    public final Vote vote;

    private CachedVote(long itemId, Type type, Vote vote) {
        this.itemId = itemId;
        this.type = type;
        this.vote = vote;
    }

    private static long voteId(Type type, long itemId) {
        return itemId * 10 + type.ordinal();
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    public static Optional<CachedVote> find(SQLiteDatabase db, Type type, long itemId) {
        try (Cursor cursor = db.rawQuery("SELECT item_id, type, vote FROM cached_vote WHERE id=?",
                new String[]{String.valueOf(voteId(type, itemId))})) {

            if (!cursor.moveToNext())
                return Optional.absent();

            return Optional.of(ofCursor(cursor));
        }
    }

    private static CachedVote ofCursor(Cursor cursor) {
        return new CachedVote(
                cursor.getLong(0),
                Type.valueOf(cursor.getString(1)),
                Vote.valueOf(cursor.getString(2)));
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    public static List<CachedVote> find(SQLiteDatabase db, Type type, List<Long> ids) {
        if (ids.isEmpty())
            return Collections.emptyList();

        List<Long> sortedIds = Ordering.natural().immutableSortedCopy(ids);
        String encodedIds = Joiner.on(",").join(transform(sortedIds, id -> voteId(type, id)));
        try (Cursor cursor = db.rawQuery(
                "SELECT item_id, type, vote FROM cached_vote WHERE id IN (" + encodedIds + ")",
                new String[]{})) {

            List<CachedVote> result = new ArrayList<>();
            while (cursor.moveToNext()) {
                result.add(ofCursor(cursor));
            }

            return result;
        }
    }


    public static void quickSave(SQLiteDatabase database, Type type, long itemId, Vote vote) {
        String stmt = String.format(Locale.ROOT,
                "INSERT OR REPLACE INTO cached_vote (id, item_id, type, vote) VALUES (%d, %d, \"%s\", \"%s\")",
                voteId(type, itemId), itemId, type.name(), vote.name());

        database.execSQL(stmt);
    }

    public static void clear(SQLiteDatabase database) {
        database.execSQL("DELETE FROM cached_vote");
    }

    public static void prepareDatabase(SQLiteDatabase db) {
        logger.info("drop cached_vote__uid index, if it exists");
        db.execSQL("DROP INDEX IF EXISTS cached_vote__uid");

        logger.info("create cached_vote table if it does not exist.");
        db.execSQL("CREATE TABLE IF NOT EXISTS cached_vote (" +
                "id INTEGER PRIMARY KEY," +
                "type TEXT," +
                "vote TEXT," +
                "item_id INTEGER)");
    }

    public enum Type {
        ITEM, COMMENT, TAG
    }
}

File: app/src/main/java/com/pr0gramm/app/orm/CachedVote.kt
package com.pr0gramm.app.orm

import app.cash.sqldelight.coroutines.asFlow
import app.cash.sqldelight.coroutines.mapToList
import app.cash.sqldelight.coroutines.mapToOneOrDefault
import com.pr0gramm.app.Logger
import com.pr0gramm.app.db.CachedVoteQueries
import com.pr0gramm.app.time
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flowOf
import java.util.EnumMap

/**
 * A cached vote.
 */
data class CachedVote(val itemId: Long, val type: Type, val vote: Vote) {
    enum class Type {
        ITEM, COMMENT, TAG
    }

    companion object {
        private val logger = Logger("CachedVote")
        private val allTypes = Type.values()

        private fun voteId(type: Type, itemId: Long): Long {
            return itemId * 10 + type.ordinal
        }

        private fun toCachedVote(itemId: Long, itemType: Int, voteValue: Int): CachedVote {
            return CachedVote(itemId, type = allTypes[itemType], vote = Vote.ofVoteValue(voteValue))
        }

        fun find(cv: CachedVoteQueries, type: Type, itemId: Long): Flow<CachedVote> {
            return cv.findOne(voteId(type, itemId), this::toCachedVote)
                .asFlow()
                .mapToOneOrDefault(CachedVote(itemId, type, Vote.NEUTRAL), Dispatchers.IO)
        }

        fun find(cv: CachedVoteQueries, type: Type, ids: List<Long>): Flow<List<CachedVote>> {
            if (ids.isEmpty()) {
                return flowOf(listOf())
            }

            // lookup votes in chunks, as sqlite can check at most 1000 parameters at once.
            val flows: List<Flow<List<CachedVote>>> = ids.chunked(512)
                .map { chunk -> chunk.map { voteId(type, it) } }
                .map { chunk -> cv.findSome(chunk, this::toCachedVote) }
                .map { it.asFlow().mapToList(Dispatchers.IO) }

            return combine(flows) { votes -> votes.asList().flatten() }
        }

        fun quickSave(cv: CachedVoteQueries, type: Type, itemId: Long, vote: Vote) {
            cv.saveVote(voteId(type, itemId), itemId, type.ordinal, vote.voteValue)
        }

        fun clear(cv: CachedVoteQueries) {
            cv.deleteAll()
        }

        fun count(cv: CachedVoteQueries): Map<Type, Map<Vote, Int>> {
            val result = HashMap<Type, EnumMap<Vote, Int>>()

            logger.time("Counting number of votes") {
                cv.count().executeAsList().forEach { count ->
                    val type = allTypes[count.itemType]
                    val vote = Vote.ofVoteValue(count.voteValue)

                    val votes = result.getOrPut(type, { EnumMap<Vote, Int>(Vote::class.java) })
                    votes[vote] = (votes[vote] ?: 0) + count.count.toInt()
                }
            }

            return result
        }
    }
}
